src/map/storage.c

```diff
@@ -356,12 +380,24 @@ int storage_storagegettocart(struct map_session_data* sd, int index, int amount)
  *------------------------------------------*/
 void storage_storageclose(struct map_session_data* sd)
 {
+	int i = 0;
+
 	nullpo_retv(sd);
 
+	Assert_retv(sd->state.storage_flag == STORAGE_FLAG_NORMAL);
+
 	clif->storageclose(sd);
 
-	if( map->save_settings&4 )
-		chrif->save(sd,0); //Invokes the storage saving as well.
+	if (map->save_settings & 4)
+		chrif->save(sd, 0); //Invokes the storage saving as well.
+
+	/* Erase deleted account storage items from memory
+	 * and resize the vector. */
+	for (i = 0; i < VECTOR_LENGTH(sd->storage); i++) {
+		if (VECTOR_INDEX(sd->storage, i).nameid == 0) {
```

Comment:
This would skip adjacent empty slots. You need to `i--` whenever you erase an entry (or if you prefer, use a while loop and only do `i++` when you haven't deleted anything)